### PR TITLE
feat(DEV-1825): self-serving royalties

### PR DESCRIPTION
**Pull request by sneljo1**
Adds RoyaltyRegistry to be able to fetch, set and clear token royalties

### CODE REVIEW

File: src/refinable/RoyaltyRegistry.ts [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+    this.contract = contract;
+    return contract;
+  }
+
+  async getRoyaltyInfo(contractAddress: string) {
+    const contract = await this._lazyGetContract();
+
+    const royaltyInfo = await contract
+      .toEthersContract(this.provider)
+      .getRoyaltyInfoForToken(contractAddress, "1");
```
Comment (viet10101988) [inline src/refinable/RoyaltyRegistry.ts:40]: is it safe? token 1 might not exist?

File: src/refinable/client/RefinableEvmClient.ts [hunk 2]
```diff
@@ -63,6 +64,10 @@ export class RefinableEvmClient {
     return new NFTBuilder(this.refinableClient, params);
   }
 
+  royaltyRegistry(chainId: Chain) {
+    return new RoyaltyRegistry(this.refinableClient, chainId);
```
Comment (viet10101988) [inline src/refinable/client/RefinableEvmClient.ts:68]: maybe this could be cached (in a map?)

**Review by viet10101988 — APPROVED** [review]

File: src/refinable/RoyaltyRegistry.ts [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+        account: royalty.account,
+        value: royalty.value.toNumber(),
+      })),
+    };
+  }
+
+  async setRoyaltiesByToken(contractAddress: string, royalties: LibPart[]) {
+    const contract = await this._lazyGetContract();
+
+    console.log({ contractAddress, royalties });
```
Comment (mihaiberq) [inline src/refinable/RoyaltyRegistry.ts:56]: log here

File: src/refinable/RoyaltyRegistry.ts [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+
+    return await contract
+      .toEthersContract()
+      .setRoyaltiesByToken(contractAddress, royalties);
+  }
+
+  async clearRoyaltiesType(contractAddress: string) {
+    const contract = await this._lazyGetContract();
+
+    await contract.toEthersContract().clearRoyaltiesType(contractAddress);
```
Comment (mihaiberq) [inline src/refinable/RoyaltyRegistry.ts:66]: im guessing this is on the contract and triggers a tx?

**Review by mihaiberq — APPROVED** [review]
im approving since it's a log that can quickly be removed before merge

**Review by colthreepv — CHANGES_REQUESTED** [review]
viet asks about tokenId, that is a good point, to retrieve a valid one should we check the supply amount? (especially delicate in erc1155)

method `getRoyaltyInfoForToken` is on the [feat/DEV-1825-self-serve-royalties](https://github.com/refinableco/nft-contracts/tree/feat/DEV-1825-self-serve-royalties), but that repo has no PR open yet

File: src/refinable/RoyaltyRegistry.ts (hunk 1 above)
Comment (sneljo1) [reply src/refinable/RoyaltyRegistry.ts:40]: Yes, it doesn't matter. At least not for the creator suite settings.

File: src/refinable/RoyaltyRegistry.ts (hunk 4 above)
Comment (sneljo1) [reply src/refinable/RoyaltyRegistry.ts:66]: Yes

**Comment by sneljo1** [comment]
> viet asks about tokenId, that is a good point, to retrieve a valid one should we check the supply amount? (especially delicate in erc1155)
> 
> method `getRoyaltyInfoForToken` is on the [feat/DEV-1825-self-serve-royalties](https://github.com/refinableco/nft-contracts/tree/feat/DEV-1825-self-serve-royalties), but that repo has no PR open yet

Retrieving a valid one is not that important for the creator suite. Strategies like the EIP standard are collection based, not per token. I updated it so that tokenId can be passed. I'll need to it fetch it for the item pages.

**Pull request merged**